src/tree/identity_tree.rs

```diff
@@ -0,0 +1,592 @@
+        Some(self.cmp(other))
+    }
+}
+
+impl std::hash::Hash for Root {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.hash.hash(state);
+    }
+}
+
+pub struct IdentityTree {
+    pub tree: DynamicMerkleTree<PoseidonHash>,
+    pub tree_updates: BTreeMap<Root, StorageUpdates>,
+    pub leaves: HashMap<Hash, u32>,
+}
+
+impl IdentityTree {
+    pub fn new(tree_depth: usize) -> Self {
+        let tree = DynamicMerkleTree::new((), tree_depth, &Hash::ZERO);
+
+        Self {
+            tree,
+            tree_updates: BTreeMap::new(),
+            leaves: HashMap::new(),
+        }
+    }
+
+    pub fn inclusion_proof(
+        &self,
+        leaf: Hash,
+        root: Option<&Root>,
+    ) -> eyre::Result<Option<InclusionProof>> {
+        let leaf_idx = self.leaves.get(&leaf).ok_or_eyre("Leaf not found")?;
+
+        if let Some(root) = root {
+            if root.hash == self.tree.root() {
+                let proof = self.tree.proof(*leaf_idx as usize);
+                Ok(Some(InclusionProof::new(self.tree.root(), proof)))
+            } else {
+                let proof = self.construct_proof_from_root(*leaf_idx, root)?;
+                Ok(Some(InclusionProof::new(root.hash, proof)))
+            }
+        } else {
+            let proof = self.tree.proof(*leaf_idx as usize);
+            Ok(Some(InclusionProof::new(self.tree.root(), proof)))
+        }
+    }
+
+    pub fn construct_proof_from_root(
+        &self,
+        leaf_idx: u32,
+        root: &Root,
+    ) -> eyre::Result<Proof> {
+        let updates = self
+            .tree_updates
+            .get(root)
+            .ok_or_eyre("Could not find root in tree updates")?;
+
+        let mut node_idx = leaf_to_storage_idx(leaf_idx, self.tree.depth());
+
+        let mut proof: Vec<Branch<Hash>> = vec![];
+
+        while node_idx > 0 {
+            let sibling_idx = if node_idx % 2 == 0 {
+                node_idx - 1
+            } else {
+                node_idx + 1
+            };
+
+            let sibling = updates
+                .get(&sibling_idx)
+                .copied()
+                .or_else(|| {
+                    let (depth, offset) =
+                        storage_idx_to_coords(sibling_idx as usize);
+                    Some(self.tree.get_node(depth, offset))
+                })
+                .expect("Could not find node in tree");
+
+            proof.push(if node_idx % 2 == 0 {
+                Branch::Right(sibling)
+            } else {
+                Branch::Left(sibling)
+            });
+
+            node_idx = (node_idx - 1) / 2;
+        }
+
+        Ok(semaphore::merkle_tree::Proof(proof))
+    }
+
+    pub fn insert(&mut self, index: u32, value: Hash) {
+        //TODO: do we want to return an error if a non unique leaf is being inserted?
+        self.leaves.insert(value, index);
+        // We can expect here because the `reallocate` implementation for Vec<H::Hash> as DynamicTreeStorage does not fail
```

Comment:
Hmmm. Yeah the return type here should probably be dependent on the underlying storage. I'll look into cleaning that up on my side.